### ISSUE TITLE
Match requests with responses/exceptions

### ISSUE THREAD

**Issue by sagikazarmark** [php-http/mock-client#3]
This way the response/exception given cannot be tied to the request. Using a similar approach to the [BatchResult](https://github.com/php-http/client-common/blob/master/src/BatchResult.php) we could.

**Comment by sagikazarmark** [comment]
Well, client common is a dependency, so you could even use BatchResult.

**Comment by ddeboer** [comment]
Will add this in a separate PR.

**Comment by ddeboer** [comment]
I agree that linking the request with the expected exception or response is useful. However, this doesn’t work in [tests where you don’t yet have a request object](https://github.com/FriendsOfSymfony/FOSHttpCache/blob/adjust-php-http/tests/Unit/ProxyClient/AbstractProxyClientTest.php#L44) because it will be created later or in another layer of the application code.

**Comment by sagikazarmark** [comment]
Tough question. Would be nice to have both behavior. Let's stick with the current one for now. We can add it if we find a use case (like `addExceptionFor` with a second, request argument). Generally it should be predictible what goes into the mock client, so no linking between requests and results should not be a problem.

**Comment by dbu** [comment]
for unit tests, the mock client should be enough. the biggest issue i currently see is that exceptions are always preferred over responses. if i want to test a flow that first returns a response but then runs into an exception, that will be impossible currently.

if we want to get more fancy, we could have something like request matchers (i think that is only a symfony concept) or callable associated with the reaction (response/exception). the mock client would go through that list and return the reaction of the first match. that would however be quite a change to the current behaviour.

**Comment by sagikazarmark** [comment]
Thinking about the following:

`$mockClient->on($request[, $requestMatcher])->return($response/$exception);`

When there is no request matcher, an ExactMatcher should be used which compares the requests based on attributes.

**Comment by dbu** [comment]
that sounds nice. i would allow to pass null für $request to say the
request is not specified. otherwise you would need a AlwaysMatch and a
request that is not relevant which feels weird.

i fear the request matcher could become quite a thing, how to define,
how to configure. must have header, must not have header, header value
must be ... and so on. or we just use callback functions and document a
few simple examples.

**Comment by sagikazarmark** [comment]
> otherwise you would need a AlwaysMatch and a request that is not relevant which feels weird.

Hm, I would probably implement AlwaysMatch anyway to be able to use the same logic (request matcher).

I would also preserve the current addResponse addException methods which would be equivalent to $request =null.

> i fear the request matcher could become quite a thing

You mean the exact matcher or request matchers in general?

I think the exact matcher would not be that hard: check method, url, other single values. Compare arrays. Optionally compare bodies?

**Comment by dbu** [comment]
> I would also preserve the current addResponse addException methods which would be equivalent to > $request =null.

that actually makes sense. so do addResponse / addException leads to returning/throwing them in the order added in the future? i think we should start preserve the order, even if that is a BC break strictly speaking. we did not release yet so its ok :-) and it only affects tests and only on the edge case that you wanted an exception, then a response and set up the mock client in the opposite order...

my fear is request matchers in general, not exact match in particular. how would you do the matchers? classes? in this repo? its actually something relevant for other use cases too.

**Comment by sagikazarmark** [comment]
> i think we should start preserve the order
Agreed. But if we use request matchers, I am not completely sure how we can do that. Using request matchers the obvious logic is a first-match logic

> my fear is request matchers in general, not exact match in particular. how would you do the matchers? classes? in this repo? its actually something relevant for other use cases too.

I think most of the general matchers can go into the message package. We can also have a callable matcher. Alternatively we can have named request matching:

`$mockClient->onMatchingHeader('Authorization', 'blah')->return()`

or even

`$mockClient->onSomething()->onSomethingElse()->return()`

to chain matching rules.

**Comment by dbu** [comment]
this is what symfony does (for its requests, not psr 7) https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/RequestMatcherInterface.php
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/RequestMatcher.php

we could port those to psr 7 (and give the implementation a fluent interface). i did a quick google search, it seems to not yet exist. but maybe its just not found by google. before we do it, we should ask around if we missed it...

**Comment by sagikazarmark** [comment]
> we could port those to psr 7

?? We already have some request matchers in the message package. You ported the one from symfony.

**Comment by joelwurtz** [comment]
I would propose to do the following : 

 * Create a `AlwaysFailClient` which it's constructed with an Http `Exception` and will always throw this exception on sendRequest or return a HttpRejectedPromise on sendAsyncRequest
 * Create a `AlwaysSucessClient` which it's constructed with an PSR ResponseInterface and will always return this response on sendRequest or return a HttpFulfilledPromise on sendAsyncRequest
 
Then we could update or MockClient which will extends the `HttpClientRouter`.

It will also have the method proposed by @sagikazarmark for a better api

```
/**
 * @param RequestInterface                       $request
 * @param ResponseInterface|Exception $return
 */
public function on(RequestInterface $request, $return);
```

This method will internally create a request matcher based on the request and associate it with a `AlwaysFailClient` or `AlwaysSuccessClient`

Additionally this client would be constructed with a defaultClient (which will be decorated by the HttpFlexibleClient to emulate sync or async call if needed) which will be used as the default client if a request does not match anything

Then for simple use case, people will use the `AlwaysFailClient` or `AlwaysSucessClient` as most of the time there is no need to do multiple http calls and API maintainer can provide a more complex and complete client by giving a `MockClient` with many matchers.

WDYT ?

**Comment by sagikazarmark** [comment]
I like the ClientRouter idea, although I would prefer composition.

Also, this would not allow us utilizing any ordered stuff, it would still require custom logic.

**Comment by joelwurtz** [comment]
> I like the ClientRouter idea, although I would prefer composition.

:+1: 

We can use the client router internally, add another method for using request matcher (or change signature, but i'm not a fan of signatures with too many possibilities)

Also we can catch the exception returned by the router to use the default client if needed (but require https://github.com/php-http/client-common/issues/44 to not catch an exception that we should actually throw)

**Comment by joelwurtz** [comment]
Also we can provide many helpers method which allows to do the following :

```
$client->onGet('/my-url')->return(200, ['Content-Type' => 'application/json'], json_encode(['foo' => 'bar']));
$client->onPost(
...
```

**Comment by dbu** [comment]
i would keep this simple. just having `on` with RequestMatcherInterface seems  enough to me. passing a request and building the matcher from that seems tricky, its hard to know which parts of the request matter or don't matter.

then i see 2 strategies: 
a) always return / throw the first match found. throw an error when no matcher matches. you can use a matcher that matches everything if you want a default response. 
b) require the first matcher to match, then pop that item and do the result.

the currently existing mock client is neither of those: a) does not remove things when they are used.
b) is more strict situation, requests are validated to come in the right order, and extra requests lead to an error.

i think these are enough scenarios and we should not need more complicated solutions like a) but with default response, or b) but default response when all responses are used up.

**Comment by Soullivaneuh** [comment]
I wrote a quite dirty and simple alternative for this issue:

```php
use Http\Mock\Client;
use Psr\Http\Message\RequestInterface;
use Psr\Http\Message\ResponseInterface;

/**
 * @author Sullivan Senechal <soullivaneuh@gmail.com>
 */
final class MockClient extends Client
{
    /**
     * @var ResponseInterface|null
     */
    private $defaultResponse;

    /**
     * @var array[]
     */
    private $responseMap = [];

    /**
     * @param ResponseInterface|null $defaultResponse
     */
    public function setDefaultResponse(?ResponseInterface $defaultResponse)
    {
        $this->defaultResponse = $defaultResponse;
    }

    /**
     * @param RequestInterface $request
     * @param ResponseInterface $response
     */
    public function addResponseMap(RequestInterface $request, ResponseInterface $response)
    {
        array_push($this->responseMap, [$request, $response]);
    }

    /**
     * {@inheritdoc}
     */
    public function sendRequest(RequestInterface $request)
    {
        foreach ($this->responseMap as $m => $mapElement) {
            /** @var RequestInterface $requestElement */
            /** @var ResponseInterface $responseElement */
            list($requestElement, $responseElement) = $mapElement;

            if ($requestElement->getRequestTarget() === $request->getRequestTarget()
                && $requestElement->getMethod() === $request->getMethod()
                && (string) $requestElement->getUri() === (string) $request->getUri()
            ) {
                unset($this->responseMap[$m]);
                return $responseElement;
            }
        }

        if ($this->defaultResponse) {
            return $this->defaultResponse;
        }

        return parent::sendRequest($request);
    }
}
```

Usage (as a Symfony service):

```php
$httplugMockClient = $this->getContainer()->get('httplug.client.mock');
$httplugMockClient->setDefaultResponse(new JsonResponse([]));
$httplugMockClient->addResponseMap(
    new Request('https://api.github.com/user', 'GET'),
    new JsonResponse([
        'name' => 'John Doe',
        'email' => 'john@doe.com',
    ])
);
```

It's working, and I think this base might be used with your `on()` system.

**Comment by adamquaile** [comment]
This would be useful for something I'm working on at the moment. If there's consensus about what the interface should look like, I'd be happy to submit a PR.

**Comment by dbu** [comment]
fixed in #27
